main/src/ca/uwaterloo/flix/language/phase/TreeShaker.scala

```diff
@@ -18,29 +18,207 @@ package ca.uwaterloo.flix.language.phase
+      case Expression.Existential(fparam, exp, loc) => visitExp(exp)
+      case Expression.Universal(fparam, exp, loc) => visitExp(exp)
+      case _ =>
+    }
+
+    /**
+      * Adds the function `sym` to the map of reachable functions.
+      */
+    def newDefinitionSymbol(sym: Symbol.DefnSym): Unit = {
+      // If `sym` has not already been determined reachable, look up its definition in `root`.
+      if (!reachableFunctions.contains(sym)) {
+        root.definitions.get(sym) match {
+          case Some(defn) =>
+            reachableFunctions.put(sym, defn)
+            queue.enqueue(defn)
+          // If `sym` is not defined in `root`, leave this for error checking later.
+          case _ =>
+        }
+      }
+    }
+
+    /*
+     * We can now use these helper functions to perform tree shaking.
+     */
+
+    // Start the timer.
+    val t = System.nanoTime()
+
+    /*
+     * Find reachable functions that:
+     *
+     * (a) Appear in the global namespaces, take zero arguments, and are not marked as synthetic.
+     */
+    for ((sym, defn) <- root.definitions) {
+      if (isReachableRoot(defn)) {
+        queue.enqueue(defn)
+        reachableFunctions.put(sym, defn)
+      }
+    }
+
+    /*
+     * Find reachable functions that:
+     *
+     * (b) Appear in a fact or a rule as a filter/transfer function.
+     */
+    for (stratum <- root.strata) {
+      for (constraint <- stratum.constraints) {
+        constraint.head match {
+          case SimplifiedAst.Predicate.Head.Positive(sym, terms, loc) => terms.foreach(visitTermHead)
+          case SimplifiedAst.Predicate.Head.Negative(sym, terms, loc) => terms.foreach(visitTermHead)
+          case _ =>
+        }
+
+        constraint.body.foreach {
+          case SimplifiedAst.Predicate.Body.Positive(sym, terms, loc) => terms.foreach(visitTermBody)
+          case SimplifiedAst.Predicate.Body.Negative(sym, terms, loc) => terms.foreach(visitTermBody)
+          case SimplifiedAst.Predicate.Body.Filter(sym, terms, loc) => newDefinitionSymbol(sym)
+          case SimplifiedAst.Predicate.Body.Loop(sym, term, loc) => visitTermHead(term)
+        }
+      }
+    }
+
+    /*
+     * Find reachable functions that:
+     *
+     * (c) Appear in a lattice declaration.
+     */
+    root.lattices.values.foreach({
+      case SimplifiedAst.Definition.Lattice(tpe, bot, top, leq, lub, glb, loc) =>
+        visitExp(bot)
+        visitExp(top)
+        visitExp(leq)
+        visitExp(lub)
+        visitExp(glb)
+    })
+
+    /*
+     * Find reachable functions that:
+     *
+     * (d) Appear in a function which itself is reachable.
+     */
+    while (queue.nonEmpty) {
+      // Extract a function body from the queue and search for other reachable functions.
+      visitExp(queue.dequeue().exp)
+    }
+
+    // Calculate the elapsed time.
+    val e = System.nanoTime() - t
 
-    root.toSuccess
+    // Reassemble the AST.
+    root.copy(
```

Comment:
Once you have the set of reachable symbols, you can just do root.definitions.filter at this point.